rfd/0164-scoped-rbac.md

```diff
@@ -0,0 +1,956 @@
+
+In our example `T2 > T1 > Tr` and `T2 > T1 > Tg`. There are two sub cases when `Tr > Tg` and `Tr < Tg`
+
+Here are possible combinations of versions that The Proxy can encounter:
+
+```
+1. G - Tg, R - Tr
+2. G - T1, R - Tr
+3. G - Tg, R - T2
+4. G - T1, R - T2
+```
+
+Let's review the cases in the reverse order:
+
+* In the cases 3 and 4, the Proxy sees that the resource group has been updated after the grant, because `T2 > T1 > Tg` and can proceed.
+* In the case 2, The proxy sees that `R` may be stale, because it's `Tr` less than `T1 and has to make a call to the backend making sure it fetched update of `R` least as fresh as `T1` with membership of `luna`.
+* In case 1, if `Tr > Tg`, the proxy can proceed, otherwise if `Tr < Tg`, the proxy has to make sure it has fetched the latest update of `Tr` at least as fresh as `Tg` with membership of `luna`.
+
+How would Teleport "make sure that it has fetched a version of a resource at least as fresh as a timestamp?". 
+
+Some backends make it easy, like Etcd giving a generation version for every resource, making sure you are not looking at the stale snapshot. 
+
+Backends like DynamoDB are making it very hard or impossible, the only way with those backends is for the client to keep track when was the last time if has last fetched the resource `R` in cases 1 and 2, and update this value for `R` in it's in memory-cache.
+
+We also are making assumption that time is in sync between all clients. In real-world, it's more complicated, because we have to consider the times when roles have been updated and fetched, but the algorithm will remain the same.
+
+TODO: question for a scale team on implementation details for backends.
+
+### Migration
+
+We will release a new versions of resources in multiple phases:
+
+* `RoleV8` will introduce `parent_resource_group` and `grantable_scope`, `create_time`, `update_time`. `RoleV9` will remove `node_labels`,  `node_labels_expression`. Users can gradually migrate to new versions, first by leveraging scopes and resoruce groups, and then removing labels matching. For new `RoleV8` roles by default, if the `grantable_scope` is missing, we assume empty scope - that will prevent the role from being granted on any scopes. When migrating existing roles from `V7`, we would set `/` - root scope to avoid breaking the cluster. 
+* Access List `V2` will also introduce `create_time` and `update_time`,  `scope` and `member_lists`.
+* We will craete a new resrouce `ResourceGroup` referenced in this RFD `V1` will have `create_time` and `update_time`.
+* The connector resources `saml` and `oidc` will loose `attributes_to_roles` and `claims_to_roles` respectively in their `V3`.
+* We will deprecate login rules in the first major release of 2025.
+* We will stop encoding roles and traits in certificatees in the first major release of 2025.
```

Comment:
I definitely welcome this change but thinking in terms of Machine ID, I see a hitch with this.

Currently, `tbot` leverages role impersonation. This creates a scenario where the roles included in the certificate do not match the roles that the user directly holds. The Bot User has a Bot Role which grants a few permissions required for `tbot` to run (e.g the ability to query the public certs of the CAs) and grants the ability to impersonate the roles that the user wants the Bot to include in outputted certificates.

I see two options as to how we handle this:

1.
We retain the ability to specify roles in the certificates explicitly for this use-case. This involves less re-engineering but means that we have to continue to support roles within certificates which is probably going to be burdensome going forward.

2.
We re-engineer Machine ID to eliminate the reliance on Role Impersonation. My intuition and personal experience with users configurations suggests that very few, if any, folks are leveraging the ability within `tbot` to produce certificates that are a subset of the roles available for the Bot User to impersonate. I imagine the primary usecase of this ability to use a subset of the roles would be useful in cases where a single `tbot` installation is producing and distributing credentials to a number of consumers which require different permissions - however - this is a case we want to discourage anyway. Running `tbot` closer to the actual service that will consume the credentials is far more secure.

We'd also need a way to ensure that the permissions that `tbot` itself needs to function (e.g the ability to query the public certs of the CAs) are not available to those consuming the credentials output by `tbot`. One thing I've been considering for a few months is implicitly providing all users of kind Bot access to these permissions. These are generally only one or two low risk permissions and removing them from the Bot Role would prevent `tbot` from functioning anyway.

Personally, I feel like pursuing number 2 reduces the long term burden and aligns with some of the other work we've been considering for Machine ID.

Reply:
@strideynet so as I understand it, you are proposing to:

* Deprecate impersonate verb, and remove it from role `v8`
* Create a new type of service `machine_service` that will consume credentials and will be bound to specific bots bound to those credentials.

Is that a fair summary?

Reply:
> Deprecate impersonate verb, and remove it from role v8

Specifically, Role Impersonation is what would be incompatible with your proposed changes and is what we'd need to remove. I think User Impersonation would be fine - but we could also consider removing it to keep things simpler in the longterm if we can determine it isn't useful.

> Create a new type of service machine_service that will consume credentials and will be bound to specific bots bound to those credentials.

I'm a little unsure what you mean by "service" here. Perhaps I wasn't super clear, but essentially I would like all Bots to have access to certain low-privilege RPCs by virtue of being a Bot type user (e.g reading the CAs). This removes the need for Role Impersonation in `tbot`.

Reply:
> Specifically, Role Impersonation is what would be incompatible with your proposed changes and is what we'd need to remove. I think User Impersonation would be fine - but we could also consider removing it to keep things simpler in the longterm if we can determine it isn't useful.

Role impersonation is a prerequisite for user impersonation. If a user's roles are not fully outlined in the impersonated roles, it becomes impossible to impersonate that user. Implementing this in TAG is a complex task.

Removing role impersonation while retaining user impersonation would constitute a behavior-altering change with potential disruptions. In my view, opting to eliminate impersonation altogether is a valid stance, but it carries the risk of disrupting numerous setups that heavily depend on it. These setups include scenarios where administrators test user permissions or generate long-standing credentials.

Reply:
> scenarios where administrators test user permissions 

Long term, it'd be nicer if we solved this through some sort of permission tester UI or CLI tool that can evaluate an arbitrary decision e.g "Can user X connect to server Y as principal Z? Yes - but they will need to use Per-session MFA and the session requires a moderator."

> generate long-standing credentials

Whilst this certainly used to be the case, I think this is becoming less and less viable - especially with the most recent changes introducing admin RPC MFA requirements to non-bots. I do agree this would be highly disruptive though

---

Overall, I'm not massively in favour of deprecating impersonation - but if we choose to move forward with the RFC as is, then we need to determine how role/user impersonation will work under this design.

Reply:
@strideynet can you give me an example of how user impersonation / role impersonation will be incompatible with this proposal?

Reply:
On a second re-read of the RFD, I think I understand now. Originally, I had missed the fact that `grant` contains a roles field and I saw this part here which mentioned that roles would be removed from the certificate. This left me wondering how we would produce an identity which had different roles to the users directly held roles - but I can see how role impersonation would work within a `grant` now.